### PR TITLE
Merge two sorted lists

### PR DESCRIPTION
**Pull request by KimaciaJnr**
---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations
- [ ] The function is not called in the function file

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/merge_two_sorted_lists [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on 04/01/2025
+@author: Peter Ngugi
+"""
```
Comment (Fikre-Mamo) [inline solutions/merge_two_sorted_lists]: Docstring for the program, need to add more details

File: solutions/merge_two_sorted_lists [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+
+def merge_two_sorted_lists(l1: list, l2: list) -> list:
+    """
+    Merges two sorted Python lists and returns a single sorted list.
+
+    Parameters:
+        l1 (list): The first sorted list.
+        l2 (list): The second sorted list.
```
Comment (Fikre-Mamo) [inline solutions/merge_two_sorted_lists]: l1 and l2 are good but it will be better to change them to meaningful parameters,
like list1 and list2

File: solutions/merge_two_sorted_lists [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    >>> merge_two_sorted_lists([], [4, 77, 90])
+    [4, 77, 90]
+
+    >>> merge_two_sorted_lists([], [])
+    []
+    """
+    if not isinstance(l1, list) or not isinstance(l2, list):
+        raise TypeError("Both inputs must be lists.")
+
```
Comment (Fikre-Mamo) [inline solutions/merge_two_sorted_lists]: `raise TypeError(f"Both inputs must be lists, got {type(l1).__name__} and {type(l2).__name__}.")` 
If one or both aren’t lists, it stops and gives an error message:
"You can only give me lists! But you gave me a [something else]."

The part {type(11).__name__} tells the program to show what kind of  data type was actually given instead of a list, like saying, `"You gave me a string and a list"`

File: solutions/merge_two_sorted_lists [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            i += 1
+        else:
+            merged.append(l2[j])
+            j += 1
+
+    # Append any remaining elements
+    merged.extend(l1[i:])
+    merged.extend(l2[j:])
+
+    return merged
```
Comment (Fikre-Mamo) [inline solutions/merge_two_sorted_lists]: Add docstring to explain this part of code

File: solutions/tests/test_merge_two_sorted_lists [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    """Tests for the merge_two_sorted_lists function"""
+
+    def test_two_single_digit_lists(self):
+        """Should return a single sorted list merged from the two single-digit lists"""
+        actual = merge_two_sorted_lists([1, 3, 5], [2, 4, 6])
+        expected = [1, 2, 3, 4, 5, 6]
+        self.assertEqual(actual, expected)
+
+    def test_one_list_empty(self):
+        """Should return a list containing values of the single list provided"""
```
Comment (Fikre-Mamo) [inline solutions/tests/test_merge_two_sorted_lists]: Test Naming Consistency:

Consider making test method names slightly more aligned with behavior. For example:
test_one_list_empty → test_one_empty_list

File: solutions/tests/test_merge_two_sorted_lists [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        actual = merge_two_sorted_lists([], [1, 2, 3])
+        expected = [1, 2, 3]
+        self.assertEqual(actual, expected)
+
+    def test_both_lists_empty(self):
+        """Should return an empty list"""
+        actual = merge_two_sorted_lists([], [])
+        expected = []
+        self.assertEqual(actual, expected)
```
Comment (Fikre-Mamo) [inline solutions/tests/test_merge_two_sorted_lists]: test_both_lists_empty → test_empty_lists

File: solutions/tests/test_merge_two_sorted_lists [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        expected = [1, 2, 3, 4, 5, 6]
+        self.assertEqual(actual, expected)
+
+    def test_lists_with_negatives(self):
+        """Should return a single sorted list with negative numbers"""
+        actual = merge_two_sorted_lists([-5, -2, 0], [-4, 1, 3])
+        expected = [-5, -4, -2, 0, 1, 3]
+        self.assertEqual(actual, expected)
+
+
```
Comment (Fikre-Mamo) [inline solutions/tests/test_merge_two_sorted_lists:91]: Add a test to check if the function correctly handles invalid inputs (e.g., non-list inputs) if error handling is implemented:
```python
def test_invalid_inputs(self):
        """Should raise a TypeError when inputs are not lists"""
        with self.assertRaises(TypeError):
            merge_two_sorted_lists("not a list", [1, 2, 3])
        with self.assertRaises(TypeError):
            merge_two_sorted_lists([1, 2, 3], None)
        with self.assertRaises(TypeError):
            merge_two_sorted_lists(123, [1, 2, 3])
        with self.assertRaises(TypeError):
            merge_two_sorted_lists([1, 2, 3], "string")
            ```

**Review by Fikre-Mamo — CHANGES_REQUESTED** [review]
# Solution Review

---

### **Behavior**

#### **Files**
- [x] The file name describes the function's behavior (`merge_two_sorted-lists`).
- [x] There is a module docstring in the function file.
- [x] The test file's name matches the function file name (`test_merge_two_sorted_lists.py`).
- [x] There is a module docstring in the tests file.

---

### **Unit Tests**
- [x] The test class has a helpful name in PascalCase (`TestMergeTwoLists`).
- [x] The test class has a docstring.
- Every unit test has:
  - [x] A helpful name.
  - [x] A clear docstring.
  - [x] Only one assertion.
  - [x] No logic in the unit test.
- [x] All tests pass.
- [x] There are tests for defensive assertions (checking input types).
- [x] There are tests for boundary cases (empty lists, unequal lengths, negative values).

---

### **Function Docstring**
- [x] The function's behavior is described.
- The function's arguments are described:
  - [x] Type (list).
  - [x] Purpose (two sorted lists to merge).
  - [x] Assumptions are implied but could be expanded (list contents should be sortable).
- The return value is described:
  - [x] Type (list).
  - [x] Assumptions are clear (merged and sorted).
- Defensive assertions are documented with `Raises:`
  - [x] The assertion checks for type (`TypeError`).
- [x] Include 3+ passing doctests in the docstring.

---

### **The Function**
- [x] The function's name describes its behavior.
- [x] The function's name matches the file name.
- [x] Correct type annotations are provided.
- [x] The function is not called within the function file.

---

### **Strategy**

#### **Do's**
- [x] Variable names (`i`, `j`, `merged`) are appropriate for their purpose.
- [x] Comments are clear and describe the strategy (two-pointer merging).
- [x] Code is well-organized with spacing between different stages.

#### **Don'ts**
- [x] Documentation focuses on the purpose, not implementation details.
- [x] Comments are minimal and meaningful.

---

### **Implementation**
- [x] Code passes formatting checks (standard, but check tools like Black or autopep8 if needed).
- [x] There are no Ruff linting or reasonable Pylint errors.
- [x] Variables use snake_case naming.
- [x] Variable names are clear and concise.
- [x] Code adheres to the described strategy simply and effectively.
- [x] No commented-out code or `print` statements.
- [x] Includes defensive assertions with minimal logic.

---

### **Summary and Suggestions**

This code is well-structured and adheres to best practices in terms of naming, defensive programming, and testing. It efficiently merges two sorted lists and handles various test cases well. However, there are a few areas for improvement that could enhance the robustness and clarity of the solution.

#### **Suggestions for Improvement**:

1. **Edge Case Test for Duplicate Values**:
   - Currently, the test cases cover various scenarios, but there isn't one specifically checking for **duplicate values** in the lists.
   - **Why?** Merging sorted lists with duplicate values should be handled correctly. Adding a test case where both lists have duplicate elements will help ensure that the function works as expected in these scenarios.
   - **Example Test Case**: 
     ```python
     def test_lists_with_duplicates(self):
         actual = merge_two_sorted_lists([1, 2, 2], [2, 3, 3])
         expected = [1, 2, 2, 2, 3, 3]
         self.assertEqual(actual, expected)
     ```

File: solutions/merge_two_sorted_lists (hunk 4 above)
Comment (KimaciaJnr) [reply solutions/merge_two_sorted_lists]: Adding comments in the while loop will only make the code verbose. The docstring well explains the parameters and return values.

File: solutions/tests/test_merge_two_sorted_lists (hunk 5 above)
Comment (KimaciaJnr) [reply solutions/tests/test_merge_two_sorted_lists]: Done!

File: solutions/tests/test_merge_two_sorted_lists (hunk 6 above)
Comment (KimaciaJnr) [reply solutions/tests/test_merge_two_sorted_lists]: Done!

File: solutions/tests/test_merge_two_sorted_lists (hunk 7 above)
Comment (KimaciaJnr) [reply solutions/tests/test_merge_two_sorted_lists:91]: All done!

File: solutions/merge_two_sorted_lists (hunk 3 above)
Comment (KimaciaJnr) [reply solutions/merge_two_sorted_lists]: Done!

File: solutions/merge_two_sorted_lists (hunk 2 above)
Comment (KimaciaJnr) [reply solutions/merge_two_sorted_lists]: Done!

File: solutions/merge_two_sorted_lists (hunk 1 above)
Comment (KimaciaJnr) [reply solutions/merge_two_sorted_lists]: Done!

File: solutions/tests/test_merge_two_sorted_lists [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+Unit tests for the merge_two_sorted_lists function. These tests verify the function's
+ability to merge two sorted lists, handle edge cases like empty lists, support lists
+with integers, floating-point numbers, and raise errors for invalid inputs like non-list types.
+
+"""
+
+import unittest
+
+from solutions import merge_two_sorted_lists
+
```
Comment (Fikre-Mamo) [inline solutions/tests/test_merge_two_sorted_lists]: need to change to --> `from solutions.merge_two_sorted_lists import .......` or `from ..merge_two_sorted_lists import ..... `

**Review by Fikre-Mamo — CHANGES_REQUESTED** [review]

**Review by Fikre-Mamo — APPROVED** [review]
Will done!

**Review by Fikre-Mamo — APPROVED** [review]
Will done!
